src/lib/network_pool/indexed_pool.ml

```diff
@@ -725,114 +731,307 @@ let rec add_from_gossip_exn :
          * Transaction_hash.User_command_with_valid_signature.t Sequence.t
        , Command_error.t )
        Result.t =
- fun ({ constraint_constants; consensus_constants; time_controller; _ } as t)
-     ~verify cmd0 current_nonce balance ->
+ fun ( { constraint_constants
+       ; consensus_constants
+       ; time_controller
+       ; slot_tx_end
+       ; _
+       } as t ) ~verify cmd0 current_nonce balance ->
   let open Command_error in
   let open Result.Let_syntax in
-  let unchecked_cmd =
-    match cmd0 with
-    | `Unchecked x ->
-        x
-    | `Checked x ->
-        Transaction_hash.User_command.of_checked x
-  in
-  let unchecked = Transaction_hash.User_command.data unchecked_cmd in
-  let verified () =
-    match cmd0 with
-    | `Checked x ->
-        Ok x
-    | `Unchecked _ ->
-        let%map x =
-          Result.of_option (verify unchecked) ~error:Invalid_transaction
-        in
-        Transaction_hash.(
-          User_command_with_valid_signature.make x
-            (User_command.hash unchecked_cmd))
-  in
-  let fee = User_command.fee_exn unchecked in
-  let fee_payer = User_command.fee_payer unchecked in
-  let nonce = User_command.nonce_exn unchecked in
-  (* Result errors indicate problems with the command, while assert failures
-     indicate bugs in Coda. *)
-  let%bind () = check_expiry t unchecked in
-  let%bind consumed = currency_consumed' ~constraint_constants unchecked in
-  let%bind () =
-    if User_command.check_tokens unchecked then return () else Error Bad_token
-  in
-  let%bind () =
-    (* TODO: Proper exchange rate mechanism. *)
-    let fee_token = User_command.fee_token unchecked in
-    if Token_id.(equal default) fee_token then return ()
-    else Error (Unwanted_fee_token fee_token)
+  let current_global_slot =
+    Consensus.Data.Consensus_time.(
+      to_global_slot
+        (of_time_exn ~constants:consensus_constants
+           (Block_time.now time_controller) ))
   in
-  (* C4 *)
-  match Map.find t.all_by_sender fee_payer with
-  | None ->
-      (* nothing queued for this sender *)
+  match slot_tx_end with
```

Comment:
I suggest to flatten this code a little bit.

Something like the following would fit nicely:

```
let%bind () = 
Result.ok_if_true ~error:After_slot_tx_end @@
Option.value_map ~f:Global_slot.(fun slot_tx_end' ->
current_global_slot >= slot_tx_end') ~default:true slot_tx_end in
```

Reply:
TODO for George: re-review the code of the function after flattening (now changes are inflated by increase of layerness)

Reply:
Good suggestion. In this case, the slot comparison condition needs to be inverted. Before it was the condition where it would fail, now it should be the condition where it should proceed.